keras_preprocessing/image.py

```diff
@@ -2098,13 +2101,21 @@ def __init__(self, dataframe, directory, image_data_generator,
         # Second, build an index of the images.
         self.filenames = []
         self.classes = np.zeros((self.samples,), dtype='int32')
-        filenames = _list_valid_filenames_in_directory(
-            directory,
-            white_list_formats,
-            self.split,
-            class_indices=self.class_indices,
-            follow_links=follow_links,
-            df=True)
+
+        if self.directory is not None:
+            filenames = _list_valid_filenames_in_directory(
+                directory,
+                white_list_formats,
+                self.split,
+                class_indices=self.class_indices,
+                follow_links=follow_links,
+                df=True)
+        else:
+            if not has_ext:
+                raise ValueError('has_ext cannot be set to False'
+                                 ' if directory is None.')
+            filenames = list(self.df[x_col][self.df[x_col].map(os.path.exists)])
```

Comment:
`os.path.exists` will return turn true for both directories and files, `os.path.isfile` will return true only for files,So I think it's better to use  `os.path.isfile` here.